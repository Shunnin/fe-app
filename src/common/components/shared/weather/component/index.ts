export { Temperature } from './temperature/temperature.component';
export { WeatherCard } from './weather-card/weather-card.component';
export { WeatherItem } from './weather-item/weather-item.component';
export * from './weather-icon';
