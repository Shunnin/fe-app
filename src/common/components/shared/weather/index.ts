export * from './component';
export { composeWeatherData, composeDailyForecastData } from './weather.service';
