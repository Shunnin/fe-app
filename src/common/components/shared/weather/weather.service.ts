import { isEmpty, map } from 'lodash-es';

import { WeatherData, DailyForecastData } from '../../../../models';
import { INT_ZERO } from '../../../utility/constant';
import { kelvinToCelsius, kphSpeed, getNextSevenDays } from '../../../utility';

export const composeWeatherData = (data: any): {} | WeatherData => {
  if (isEmpty(data)) {
    return {};
  }

  const {
    name,
    main: { temp, feels_like: feelsLike, temp_min: tempMin, temp_max: tempMax, pressure, humidity },
    wind: { speed, deg },
  } = data;
  const { id, description, main: forecast, icon } = data.weather[INT_ZERO];

  return {
    id,
    forecast,
    description,
    icon,
    name,
    tempDetails: {
      temp: kelvinToCelsius(temp),
      feelsLike: kelvinToCelsius(feelsLike),
      tempMin: kelvinToCelsius(tempMin),
      tempMax: kelvinToCelsius(tempMax),
      pressure,
      humidity,
    },
    wind: {
      speed: kphSpeed(speed),
      deg,
    },
  };
};

export const composeDailyForecastData = (data: any): DailyForecastData[] => {
  const next7Days = getNextSevenDays();
  const { list } = data || {};

  if (isEmpty(list)) {
    return [];
  }

  return map(list, (item: any, index: number) => {
    const {
      weather,
      temp: { max, min },
    } = item;
    const { id, main: forecast } = weather[INT_ZERO];

    return {
      day: next7Days[index],
      id,
      forecast,
      tempDetails: {
        tempMax: kelvinToCelsius(max),
        tempMin: kelvinToCelsius(min),
      },
    };
  });
};
