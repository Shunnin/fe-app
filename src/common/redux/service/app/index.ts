export { CHANGE_TEMP_UNIT, changeTempUnit } from './app.action';
export { appReducer } from './app.reducer';
export { APP_MODULE } from './app.constant';
