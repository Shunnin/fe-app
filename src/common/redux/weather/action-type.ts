export const GET_CURRENT_WEATHER_REQUEST = 'GET_CURRENT_WEATHER_REQUEST';
export const GET_CURRENT_WEATHER_SUCCESS = 'GET_CURRENT_WEATHER_SUCCESS';
export const GET_CURRENT_WEATHER_ERROR = 'GET_CURRENT_WEATHER_ERROR';

export const GET_DAILY_FORECAST_REQUEST = 'GET_DAILY_FORECAST_REQUEST';
export const GET_DAILY_FORECAST_SUCCESS = 'GET_DAILY_FORECAST_SUCCESS';
export const GET_DAILY_FORECAST_ERROR = 'GET_DAILY_FORECAST_ERROR';

export const GET_LOCATION_REQUEST = 'GET_LOCATION_REQUEST';
export const GET_LOCATION_SUCCESS = 'GET_LOCATION_SUCCESS';
export const GET_LOCATION_ERROR = 'GET_LOCATION_ERROR';
