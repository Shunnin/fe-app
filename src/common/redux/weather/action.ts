import {
  GET_CURRENT_WEATHER_REQUEST,
  GET_CURRENT_WEATHER_SUCCESS,
  GET_CURRENT_WEATHER_ERROR,
  GET_DAILY_FORECAST_REQUEST,
  GET_DAILY_FORECAST_SUCCESS,
  GET_DAILY_FORECAST_ERROR,
  GET_LOCATION_REQUEST,
  GET_LOCATION_SUCCESS,
  GET_LOCATION_ERROR,
} from './action-type';
import {
  ISearchWeatherPayload,
  ISearchWeatherErrorPayload,
  IGetCurrentWeatherSuccessPayload,
  IGetCurrentWeatherRequest,
  IGetCurrentWeatherSuccess,
  IGetCurrentWeatherError,
  IGetDailyForecastRequest,
  IGetDailyForecastSuccess,
  IGetDailyForecastSuccessPayload,
  IGetDailyForecastError,
  IGetLocationRequestPayload,
  IGetLocationSuccessPayload,
  IGetLocationRequest,
  IGetLocationSuccess,
  IGetLocationError,
} from './type';

export const getCurrentWeatherRequest = (payload: ISearchWeatherPayload): IGetCurrentWeatherRequest => ({
  type: GET_CURRENT_WEATHER_REQUEST,
  payload,
});

export const getCurrentWeatherSuccess = (payload: IGetCurrentWeatherSuccessPayload): IGetCurrentWeatherSuccess => ({
  type: GET_CURRENT_WEATHER_SUCCESS,
  payload,
});

export const getCurrentWeatherError = (payload: ISearchWeatherErrorPayload): IGetCurrentWeatherError => ({
  type: GET_CURRENT_WEATHER_ERROR,
  payload,
});

export const getDailyForecastRequest = (payload: ISearchWeatherPayload): IGetDailyForecastRequest => ({
  type: GET_DAILY_FORECAST_REQUEST,
  payload,
});

export const getDailyForecastSuccess = (payload: IGetDailyForecastSuccessPayload): IGetDailyForecastSuccess => ({
  type: GET_DAILY_FORECAST_SUCCESS,
  payload,
});

export const getDailyForecastError = (): IGetDailyForecastError => ({
  type: GET_DAILY_FORECAST_ERROR,
});

export const getLocationRequest = (payload: IGetLocationRequestPayload): IGetLocationRequest => ({
  type: GET_LOCATION_REQUEST,
  payload,
});

export const getLocationSuccess = (payload: IGetLocationSuccessPayload): IGetLocationSuccess => ({
  type: GET_LOCATION_SUCCESS,
  payload,
});

export const getLocationError = (): IGetLocationError => ({
  type: GET_LOCATION_ERROR,
});
