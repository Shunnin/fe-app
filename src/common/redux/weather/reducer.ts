import {
  GET_CURRENT_WEATHER_REQUEST,
  GET_CURRENT_WEATHER_SUCCESS,
  GET_CURRENT_WEATHER_ERROR,
  GET_DAILY_FORECAST_REQUEST,
  GET_DAILY_FORECAST_SUCCESS,
  GET_DAILY_FORECAST_ERROR,
  GET_LOCATION_SUCCESS,
  GET_LOCATION_ERROR,
} from './action-type';

import { WeatherActions, IWeatherState } from './type';

const initialState: IWeatherState = {
  loading: false,
  weather: {},
  dailyForecast: {},
  locations: [],
  errors: null,
};

export default (state = initialState, action: WeatherActions) => {
  switch (action.type) {
    case GET_CURRENT_WEATHER_REQUEST:
      return {
        ...state,
        loading: true,
      };

    case GET_CURRENT_WEATHER_SUCCESS:
      return {
        ...state,
        loading: false,
        weather: action.payload,
        errors: null,
      };

    case GET_CURRENT_WEATHER_ERROR:
      return {
        ...state,
        loading: false,
        weather: {},
        dailyForecast: {},
        errors: action.payload,
      };

    case GET_DAILY_FORECAST_REQUEST:
      return {
        ...state,
        loading: true,
      };

    case GET_DAILY_FORECAST_SUCCESS:
      return {
        ...state,
        loading: false,
        dailyForecast: action.payload,
      };

    case GET_DAILY_FORECAST_ERROR:
      return {
        ...state,
        loading: false,
        dailyForecast: [],
      };

    case GET_LOCATION_SUCCESS:
      return {
        ...state,
        locations: action.payload,
      };

    case GET_LOCATION_ERROR:
      return {
        ...state,
        locations: [],
      };

    default:
      return {
        ...state,
      };
  }
};
