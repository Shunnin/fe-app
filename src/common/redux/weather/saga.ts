import axios from 'axios';
import { all, call, put, takeLatest } from 'redux-saga/effects';
import { reduce, join } from 'lodash-es';

import {
  INT_ZERO,
  API_URL_GET_CURRENT_WEATHER,
  API_URL_GET_DAILY_FORECAST,
  API_URL_SEARCH_LOCATION,
} from '../../utility/constant';
import { UrlUtil } from '../../utility';

import {
  getCurrentWeatherError,
  getCurrentWeatherSuccess,
  getDailyForecastSuccess,
  getDailyForecastError,
  getLocationSuccess,
  getLocationError,
} from './action';
import { GET_CURRENT_WEATHER_REQUEST, GET_DAILY_FORECAST_REQUEST, GET_LOCATION_REQUEST } from './action-type';

const getCurrentWeather = (value: any) => {
  const apiUrl = UrlUtil.buildQueryUrl(API_URL_GET_CURRENT_WEATHER, {
    ...value,
    appid: '20571ab45c74dc2a1897b60c5b8047a1',
  });

  return axios.get(apiUrl);
};

const getDailyForecast = (value: any) => {
  console.log('file: saga.ts ~ line 30 ~ getDailyForecast ~ value', value);
  const apiUrl = UrlUtil.buildQueryUrl(API_URL_GET_DAILY_FORECAST, {
    ...value,
    appid: '20571ab45c74dc2a1897b60c5b8047a1',
  });

  return axios.get(apiUrl);
};

const searchLocation = (value: any) => {
  const data = JSON.stringify({
    query: value,
    type: 'city',
    language: 'en',
  });

  return axios.post(API_URL_SEARCH_LOCATION, data);
};

export interface ResponseGenerator {
  config?: any;
  data?: any;
  headers?: any;
  request?: any;
  status?: number;
  statusText?: string;
}

const getCurrentWeatherSaga = function* (action) {
  try {
    const response: ResponseGenerator = yield call(getCurrentWeather, action.payload);

    yield put(getCurrentWeatherSuccess(response.data));
  } catch (e) {
    yield put(getCurrentWeatherError(e));
  }
};

const getDailyForecastSaga = function* (action) {
  try {
    const response: ResponseGenerator = yield call(getDailyForecast, action.payload);

    yield put(getDailyForecastSuccess(response.data));
  } catch (e) {
    yield put(getDailyForecastError());
  }
};

const getLocationSaga = function* (action) {
  try {
    const response: ResponseGenerator = yield call(searchLocation, action.payload);
    const { data: { hits } = {} } = response;
    const locations = reduce(
      hits,
      (result, item) => {
        if (item.is_city) {
          result.push(join([item.locale_names[INT_ZERO], item.country], ', '));
        }

        return result;
      },
      [],
    );

    yield put(getLocationSuccess(locations));
  } catch (e) {
    yield put(getLocationError());
  }
};

const weatherSaga = function* () {
  try {
    yield all([
      takeLatest(GET_CURRENT_WEATHER_REQUEST, getCurrentWeatherSaga),
      takeLatest(GET_DAILY_FORECAST_REQUEST, getDailyForecastSaga),
      takeLatest(GET_LOCATION_REQUEST, getLocationSaga),
    ]);
  } catch (e) {
    // ...
  }
};

export default weatherSaga;
