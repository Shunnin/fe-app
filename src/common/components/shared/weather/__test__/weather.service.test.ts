import { composeWeatherData, composeDailyForecastData } from '../weather.service';

describe('Weather Service', () => {
  describe('composeWeatherData()', () => {
    it.each([
      [undefined, {}],
      [null, {}],
      [{}, {}],
      [[], {}],
    ])('given %p as arguments, returns %p', (data, expectedResult) => {
      const result = composeWeatherData(data);

      expect(result).toEqual(expectedResult);
    });

    it('return correct weather info when provide valid data', () => {
      const validData = {
        coord: { lon: 106.6667, lat: 10.75 },
        weather: [{ id: 802, main: 'Clouds', description: 'scattered clouds', icon: '03d' }],
        base: 'stations',
        main: {
          temp: 306.16,
          feels_like: 309.18,
          temp_min: 306.16,
          temp_max: 306.16,
          pressure: 1006,
          humidity: 49,
        },
        visibility: 10000,
        wind: { speed: 3.09, deg: 170, gust: 8.75 },
        clouds: { all: 40 },
        dt: 1650273535,
        sys: { type: 1, id: 9314, country: 'VN', sunrise: 1650235253, sunset: 1650279854 },
        timezone: 25200,
        id: 1566083,
        name: 'Ho Chi Minh City',
        cod: 200,
      };

      const result = composeWeatherData(validData);

      expect(result).toEqual({
        description: 'scattered clouds',
        forecast: 'Clouds',
        icon: '03d',
        id: 802,
        name: 'Ho Chi Minh City',
        tempDetails: {
          feelsLike: 36,
          humidity: 49,
          pressure: 1006,
          temp: 33,
          tempMax: 33,
          tempMin: 33,
        },
        wind: {
          deg: 170,
          speed: 11,
        },
      });
    });
  });

  describe('composeDailyForecastData()', () => {
    it.each([
      [undefined, []],
      [null, []],
      [{}, []],
      [[], []],
    ])('given %p as arguments, returns %p', (data, expectedResult) => {
      const result = composeDailyForecastData(data);

      expect(result).toEqual(expectedResult);
    });

    it('return correct weather info when provide valid data', () => {
      const validData = {
        city: {
          id: 1566083,
          name: 'Ho Chi Minh City',
          coord: { lon: 106.6667, lat: 10.75 },
          country: 'VN',
          population: 1000000,
          timezone: 25200,
        },
        cod: '200',
        message: 0.0606551,
        cnt: 7,
        list: [
          {
            dt: 1650254400,
            sunrise: 1650235253,
            sunset: 1650279854,
            temp: { day: 307.49, min: 299.73, max: 308.82, night: 300.88, eve: 305.16, morn: 299.91 },
            feels_like: { day: 310.02, night: 303.3, eve: 307.78, morn: 301.98 },
            pressure: 1010,
            humidity: 43,
            weather: [{ id: 500, main: 'Rain', description: 'light rain', icon: '10d' }],
            speed: 6.13,
            deg: 141,
            gust: 10.33,
            clouds: 60,
            pop: 0.25,
            rain: 0.29,
          },
          {
            dt: 1650340800,
            sunrise: 1650321623,
            sunset: 1650366257,
            temp: { day: 307.08, min: 299.11, max: 308.77, night: 300.96, eve: 303.99, morn: 299.44 },
            feels_like: { day: 309.53, night: 303.58, eve: 306.1, morn: 299.44 },
            pressure: 1011,
            humidity: 44,
            weather: [{ id: 500, main: 'Rain', description: 'light rain', icon: '10d' }],
            speed: 4.88,
            deg: 138,
            gust: 9.53,
            clouds: 68,
            pop: 0.43,
            rain: 1.64,
          },
          {
            dt: 1650427200,
            sunrise: 1650407994,
            sunset: 1650452661,
            temp: { day: 305.29, min: 299.4, max: 308.75, night: 301.21, eve: 304.74, morn: 299.55 },
            feels_like: { day: 308.54, night: 304.07, eve: 307.71, morn: 299.55 },
            pressure: 1011,
            humidity: 53,
            weather: [{ id: 500, main: 'Rain', description: 'light rain', icon: '10d' }],
            speed: 5.13,
            deg: 139,
            gust: 9.2,
            clouds: 97,
            pop: 0.6,
            rain: 1.9,
          },
        ],
      };

      const result = composeDailyForecastData(validData);

      expect(result).toEqual([
        {
          day: 'Wed',
          forecast: 'Rain',
          id: 500,
          tempDetails: {
            tempMax: 36,
            tempMin: 27,
          },
        },
        {
          day: 'Thu',
          forecast: 'Rain',
          id: 500,
          tempDetails: {
            tempMax: 36,
            tempMin: 26,
          },
        },
        {
          day: 'Fri',
          forecast: 'Rain',
          id: 500,
          tempDetails: {
            tempMax: 36,
            tempMin: 26,
          },
        },
      ]);
    });
  });
});
