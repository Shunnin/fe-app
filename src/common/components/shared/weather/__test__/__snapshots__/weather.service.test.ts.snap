// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Weather Service composeDailyForecastData() return correct weather info when provide valid data 1`] = `
Array [
  Object {
    "day": "Thu",
    "forecast": "Rain",
    "id": 500,
    "tempDetails": Object {
      "tempMax": 36,
      "tempMin": 27,
    },
  },
  Object {
    "day": "Fri",
    "forecast": "Rain",
    "id": 500,
    "tempDetails": Object {
      "tempMax": 36,
      "tempMin": 26,
    },
  },
  Object {
    "day": "Sat",
    "forecast": "Rain",
    "id": 500,
    "tempDetails": Object {
      "tempMax": 36,
      "tempMin": 26,
    },
  },
]
`;
